### DISCUSSION TITLE
Covariant template types in collections

### DISCUSSION THREAD

**Discussion by martin-heralecky** [laravel/framework#40889] [Ideas] [answered]
(see [answer below](https://github.com/laravel/framework/discussions/40889#discussioncomment-7756182))
~I suggest that types of values in collections are declared covariant, so that static analyzers work better with generic types and inheritance.~

PhpStan natively treats types of values of arrays (and iterables in general) as covariant, which I think is what people usually expect (`iterable<Child>` is a subtype of `iterable<Parent>` when `Child extends Parent`). This, however, is not true for Laravel's `Enumerable`, `LazyCollection` and `Collection`, because they declare the template type with `@template TValue` instead of `@template-covariant TValue`.

Here's an example, for which PhpStan fails:

```php
use Illuminate\Support\Collection;

class FooParent {}

class FooChild extends FooParent {}

interface BarInterface {
    /**
     * @return Collection<int, FooParent>    <-- notice FooParent
     */
    public function bar(): Collection;
}

class BarImpl implements BarInterface {
    public function bar(): Collection {
        return collect([
            0 => new FooChild(),          // <-- notice FooChild
        ]);
    }
}
```

```
Method BarImpl::bar() should return Illuminate\Support\Collection<int, FooParent> but returns Illuminate\Support\Collection<int, FooChild>.
```

If we replace `Collection` with `array`, the analysis passes.

Btw, Psalm supports the `@template-covariant` annotation as well (https://psalm.dev/docs/annotating_code/templated_annotations/#template-covariance).

**Comment by Radiergummi** [comment]
Ah, that's an annoying one. I wish the static analysis community had opted to do the opposite and make covariant the default behaviour, but that's what we got. I'm creating a PR for this.

**Reply by Radiergummi** [reply]
Ah, snap. Just discovered this has already been implemented in #49234, but only for Laravel >=10.

**Reply by martin-heralecky** [reply, marked as answer]
I actually don't think it's a good idea anymore. Covariance makes sense only for read-only collections (see e.g. comments on this [SO answer](https://softwareengineering.stackexchange.com/a/399279/212224) and [related article](https://learn.microsoft.com/de-de/archive/blogs/ericlippert/covariance-and-contravariance-in-c-part-two-array-covariance)).

Anyway, since it's already been implemented, I guess this can be closed.
